### ISSUE TITLE
Would you please provide us with the training code? Thanks a lot.

### ISSUE THREAD

**Issue by RayShark** [acrarshin/CPSC2019#1]
We are working on a project very similar to yours. And we encountered some bottlenecks in the research process. We are very excited to see that you have published the relevant source code, but we got some problems on the training part.

We are confused about how to design the **training loss** and how to process the **output** of the model. We'll be greatly appreciated if you can reply to us.

**Comment by acrarshin** [comment]
Hi Ray 
I have been pushing this off for way too long. Will start working on polishing the code now. Expect it to be up within a couple days. As for immediate help, pytorch has the loss nn.SmoothL1 built-in. If you can elucidate the issue you are having with the output, I can clarify it for you.

**Comment by RayShark** [comment]
I'm so happy for your replying, and I'd like to put some of my code and point out my confusion.
First I'll put my poor result... There must be something wrong.
![image](https://user-images.githubusercontent.com/26996377/81260083-deacfa00-906b-11ea-8925-3dfb3f26bbc3.png)
## 1. The dataset function:
```python
class Points2019(Dataset):
    def __init__(self, data_path, train=True):
        super(Points2019, self).__init__()
        self.ecg_path = './train/data/'
        self.ref_path = './train/ref/'
        dd = torch.load(data_path)
        self.train = train
        self.data = dd['X_train'] if train else dd['X_test']
        self.ref = dd['y_train'] if train else dd['y_test']
    def __getitem__(self, index):
        fid = self.data[index].split('_')[1]
        fs_ = 500
        ecg_data = np.transpose(sio.loadmat(osp.join(self.ecg_path, 'data_' + fid))['ecg'])[0]
        r_ref = sio.loadmat(osp.join(self.ref_path, 'R_' + fid))['R_peak'].flatten()
        r_ref = r_ref[(r_ref >= 0.5*fs_) & (r_ref <= 9.5*fs_)]
        r_mask = np.zeros(5000)
        r_mask[r_ref] = 1
        r_mask = r_mask.astype(np.float32)
        x = torch.Tensor(ecg_data).view((1, ecg_data.shape[0]))
        target = torch.tensor(r_mask, dtype=torch.float32)
        return x, target
    def __len__(self):
        return len(self.data)
```
I read the ref from the mat as the template code does. And I foud the output shape of your Unet is (N, 1, 5000), so I establish a mask matrix which shape is (5000,). The index that matches the R position will be set to 1 and others to be 0.
I wonder whether this method is correct to fit the model training~
## 2.Train_epoch Function:
```python
def train_epoch(model, optimizer, criterion, train_dataloader, show_interval=10):
    model.train()
    rec_acc_meter, hr_acc_meter, loss_meter, it_count = 0, 0, 0, 0
    for inputs, target in train_dataloader:
        inputs = inputs.to(device)
        target = target.to(device)
        # zero the parameter gradients
        optimizer.zero_grad()
        # forward
        output = model(inputs)
        output = output.view((-1, output.shape[2]))
        loss = criterion(output, target)
      
        loss.backward()
        optimizer.step()
        loss_meter += loss.item()
        rec_acc, hr_acc = get_score(target, output)
        rec_acc_meter += rec_acc
        hr_acc_meter += hr_acc
        it_count += 1
        if it_count != 0 and it_count % show_interval == 0:
            print("%d,loss:%.3e Rec_acc:%4f Hr_acc:%4f" % (it_count, loss.item(), rec_acc_meter/it_count, hr_acc_meter/it_count))
    return loss_meter / it_count, rec_acc_meter/it_count, hr_acc_meter/it_count
```
The original output shape is (BatchSize, 1, 5000) right? And I reshape it to (BatchSize, 5000) to be the same as the ref(See the above dataset fun.) 
I found some other researchers use the Unet, and I use the criterion **nn.BCEWithLogitsLoss()** as theirs. My result is very poor and I thought there must be some mistakes in those above functions...
## 3. The prediction of the model and scoring
The following is the function that I wrote for the scoring in each batch
```python
def get_score(y_true, y_pre):
    # The input data shape is (BatchSize, 5000)
    # Main parts are the same as the template entry code challenge.py
    fs_ = 500
    HR_ref = []
    R_ref = []
    HR_ans = []
    R_ans = []
    # y_pre = F.sigmoid(y_pre)
    # Get HR_ref， R_ref， HR_ans， R_ans
    for i in range(y_true.shape[0]):
        # Caculate sample by sample in each Batch
        y_true_temp = y_true[i, :]
        y_pre_temp = y_pre[i, :]
        y_true_temp = y_true_temp.cpu().detach().numpy().astype(np.int)
        y_pre_temp = y_pre_temp.cpu().detach().numpy().astype(np.float) 
        # Turn the Mask Matrix back to the R position list for scoring
        r_ref = np.array([ind for ind,x in enumerate(y_true_temp) if x==1])
        r_ans = find_peaks(y_pre_temp,distance = 90,height = -0.4,prominence = 0.035)[0] # the original variable settings
        # Is it enough to only use this find_peaks function to get the R_ans from the model output? I'm a little confused
       # The following is almost the same as the entry code
        r_ans = r_ans[(r_ans >= 0.5*fs_) & (r_ans <= 9.5*fs_)]
        
        r_hr_ans = np.array([loc for loc in r_ans if (loc > 5.5 * fs_ and loc < 5000 - 0.5 * fs_)])
        hr_ans = round( 60 * fs_ / np.mean(np.diff(r_hr_ans)))
        
        r_hr_ref = np.array([loc for loc in r_ref if (loc > 5.5 * fs_ and loc < 5000 - 0.5 * fs_)])
        hr_ref = round( 60 * fs_ / np.mean(np.diff(r_hr_ref)))
        
        HR_ref.append(hr_ref)
        R_ref.append(r_ref)
        HR_ans.append(hr_ans)
        R_ans.append(r_ans)
    FS = 500
    THR = 0.075
    rec_acc, hr_acc = score(R_ref, HR_ref, R_ans, HR_ans, FS, THR)
    return rec_acc, hr_acc
```
I mainly explained my problem in the Comment. It is my confusion about the output--Just use the find_peaks function to generate the r_ans from the output? Or there are some missing procedures or post-processing to predict the result from the model output?

It would be great if you could spare some moment to help me take a look.

Yours Sincerely,
Ray

**Comment by acrarshin** [comment]
Firstly, I would suggest you to look at this repo: https://github.com/acrarshin/RPNet, which has a paper associated with it. It's preprint is available on arxiv: https://arxiv.org/abs/2004.08103. This repo is extensive. However, the training code has not been released for that as well.

1) The reason for the poor result --> The number of R Peaks in a single ECG window is sparse. What this means is that, for 5000 samples in a window, there are only 10-40 r-peaks at max. This leads to a massive class imbalance problem. In order to alleviate this, I frame the problem as a regression problem. Instead of trying to output whether there is a peak or not, I try to predict the distance from the R-Peak. So instead of predicting a 1 or a 0 at each sample location, I try to predict the distance (no of samples) from the R-Peak location. Since it is a regression problem, I use something similar to an L1 Loss --> the smoothL1 loss.

2) I'd suggest you keep it as  (BatchSize, 1, 5000) as Conv1d expects that as the input. 

3) The scoring function seems correct. The post-processing only involves using find_peaks.

Again, I would highly recommend that you take a look at https://arxiv.org/abs/2004.08103

**Comment by RayShark** [comment]
### Thank you very much for your patient replies！
After reading the paper, I totally understand the whole algorithm and the training parts. I transform the r_ref to a distance_map and I applied the SmoothL1 loss as you suggest.
Now the results seem satisfactory after several epochs.
![image](https://user-images.githubusercontent.com/26996377/81316906-c31f0f00-90be-11ea-9c3b-589b540f3fc1.png)
Thanks again~~~
